doc/nom_recipes.md

```diff
@@ -0,0 +1,235 @@
+
+These are short recipes for accomplishing common tasks with nom.
+
+* [Whitespace](#whitespace)
+  + [Wrapper combinators that eat whitespace before and after a parser](#wrapper-combinators-that-eat-whitespace-before-and-after-a-parser)
+* [Comments](#comments)
+  + [`// C++/EOL-style comments`](#----c---eol-style-comments-)
+  + [`/* C-style comments */`](#----c-style-comments----)
+* [Identifiers](#identifiers)
+  + [`Rust-Style Identifiers`](#-rust-style-identifiers-)
+* [Literal Values](#literal-values)
+  + [Escaped Strings](#escaped-strings)
+  + [Integers](#integers)
+    - [Hexadecimal](#hexadecimal)
+    - [Octal](#octal)
+    - [Binary](#binary)
+    - [Decimal](#decimal)
+  + [Floating Point Numbers](#floating-point-numbers)
+
+## Whitespace
+
+
+
+### Wrapper combinators that eat whitespace before and after a parser
+
+```rust
+/// A combinator that takes a parser `inner` and produces a parser that also consumes both leading and 
+/// trailing whitespace, returning the output of `inner`.
+fn ws<'a, F: 'a, O, E: ParseError<&'a str>>(inner: F) -> impl Fn(&'a str) -> IResult<&'a str, O, E>
+  where
+  F: Fn(&'a str) -> IResult<&'a str, O, E>,
+{
+  move |i| {
+    delimited(
+      multispace0,
+      &inner,
+      multispace0
+    )(i)
+  }
+}
+```
+
+To eat only trailing whitespace, replace `delimited(...)` with `terminated(&inner, multispace0)`. Likewise, the eat only leading whitespace, replace `delimited(...)` with `preceded(multispace0, &inner)`. You can use your own parser instead of `multispace0` if you want to skip a different set of lexemes. 
+
+## Comments
+
+### `// C++/EOL-style comments`
+
+This version does not consume the newline character and returns an output of `()`.
+
+```rust
+pub fn peol_comment<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, (), E>
+{
+  map(pair(char('%'), is_not("\n\r")),
+      |_| () // Output is thrown away.
+  )(i)
+}
+```
+
+### `/* C-style comments */`
+
+Inline comments surrounded with sentinel tags `(*` and `*)`. This version returns an output of `()` and does not handle nested comments.
+
+```rust
+pub fn pinline_comment<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, (), E> {
+  map(
+    tuple((
+      tag("(*"),
+      take_until("*)"),
+      tag("*)")
+    )),
+    |_| () // Output is thrown away.
+  )(i)
+}
+```
+
+## Identifiers
+
+### `Rust-Style Identifiers`
+
+Parsing identifiers that may start with a letter (or underscore) and may contain underscores, letters and numbers may be parsed like this:
+
+```rust
+pub fn identifier(input: &str) -> IResult<&str, &str> {
+  recognize(
+    pair(
+      alt((alpha1, tag("_"))),
+      many0(alt((alphanumeric1, tag("_"))))
+    )
+  )(input)
+}
+```
+
+Let's say we apply this to the identifier `hello_world123abc`.
+The first `alt` parser would recognize `hello_world`. The `pair` combinator ensures that `123abc` will be piped to the next `alphanumeric0` parser, which recognizes every remaining character.
```

Comment:
The first `alt` parser actually recognizes only `hello` since it delegates to `alpha1`

Reply:
No, the first `alt` only recognizes the `'h'`. But yeah, looks like I forgot to update the text with the code.

Reply:
Here is a test that shows the first `alt` matches `hello`. Try it for yourself. This is because `alpha1` greedily matches 1 or more alphabetic characters, so it will match `h`, `e`, `l`, `l`, and `o` before failing to parse `_`.
```rust
use nom::branch::*;
use nom::bytes::complete::*;
use nom::character::complete::*;
use nom::multi::*;
use nom::sequence::*;
use nom::*;

pub fn identifier(input: &str) -> IResult<&str, (&str, Vec<&str>)> {
  pair(
    alt((alpha1, tag("_"))),
    many0(alt((alphanumeric1, tag("_"))))
  )(input)
}

assert_eq!(identifier("hello_world123abc"), Ok(("", ("hello", vec!["_", "world123abc"]))));
```

Reply:
Oh, no test necessary, as it's completely obvious! \**facepalm*\* I had edited my version of the function. *Doh!*

I appreciate the excellent example, though!